File: linked-list/remove_nth_node_from_end_of_list.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    """
    @param: head: The first node of linked list.
    @param: n: An integer
    @return: The head of linked list.
    """
    def removeNthFromEnd(self, head, n):
        # write your code here

        if n <= 0:
            return None
            
        dummy = ListNode(0)
        dummy.next = head
        
        pre = dummy
        
        for i in range(n):
            if not head:
                return None
            head = head.next
            
        while head:
            head = head.next
            pre = pre.next
            
        pre.next = pre.next.next
        return dummy.next

File: linked-list/remove_nth_node_from_end_of_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    
    var nodes = make([]*ListNode, 0)
    var length = 0
    cur := head
    
    for cur != nil {
        nodes = append(nodes, cur)
        cur = cur.Next
        length ++
    }
    
    if length <= 1 {
        return nil
    }
    
    if length == n {
        return head.Next
    }
    
    nodes = append(nodes, nil)
        
    nodes[length - n - 1].Next = nodes[length - n + 1]
    
    return head
}
